### DISCUSSION TITLE
Add JXL Support (JPEG XL)

### DISCUSSION THREAD

**Discussion by rschiefer** [SixLabors/ImageSharp#2507] [Ideas]
Apple just announced (last month) they will be support JXL very soon.  

https://en.wikipedia.org/wiki/JPEG_XL

Discussed briefly in this idea as well but wanted to separate it out - #1320

**Comment by brianpopow** [comment]
The chromium project on the other side has rejected support for JXL: https://bugs.chromium.org/p/chromium/issues/detail?id=1178058#c84

**Reply by IldarKhayrutdinov** [reply]
I recently read an article about this [unexpected decision](https://motionmill.com/2023/02/google-stopt-jpex-xl-gebruiken/), really it's sad for JXL.
Already being discussed: [Re-open JPEG XL issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1451807).

**Reply by AshtakaOOf** [reply]
A Rust JPEG XL decoder is now in both Chromium and Firefox.

**Comment by Hefaistos68** [comment]
Since other leading vendors like Adobe, Apple and Samsung are fully supporting JPGXL formats, its about time to add it to this otherwise excellent library. Who gives a damn about chromium's decision? They want to force the world to use WEBP, which isnt bad, but not the last word on image formats.

**Reply by JimBobSquarePants** [reply]
I want to do it. It's a multi-month endeavor though so cannot possibly do it alone.

**Reply by Hefaistos68** [reply]
Well, I might be able to help out. I have seen that there are some efforts on the way for .net versions or at least .net wrappers for the C++ libraries. Not sure which way you'd like to go.

**Reply by JimBobSquarePants** [reply]
That would be amazing if you could. We always go fully managed for easier deployment and deep integration with our pixel formats.

**Reply by Hefaistos68** [reply]
Well, in this case it means rewriting the whole library. Not a small task. But interesting. Lets see. I'll have a look into existing code and how it works.

**Comment by Agrarvolution** [comment]
Is there any update on the current state of this?
Chromium just recently added back JXL under a feature flag. You can display JPEG-XL images since February.
https://www.phoronix.com/news/Chrome-145-Released
Firefox seems to have restarted work on their implementation as well:
https://bugzilla.mozilla.org/show_bug.cgi?id=1539075

**Reply by JimBobSquarePants** [reply]
No. If there were you would see them in a branch.

It's a massive undertaking and will come after the v4 release. I'm not sure it's fair to expect us to match the cadence of browser companies employing thousands of developers

**Reply by Agrarvolution** [reply]
Thanks for the swift reply! This wasn't an expection at all - I'm sorry for coming accross that way. I was just curious since this project was mentioned in blocking Astro from getting out of the box JXL support. I hope you guys will get support from their sponsors in the end. 🫣

**Reply by JimBobSquarePants** [reply]
No worries. What is Astro? I'd like to see the thread.

**Reply by Agrarvolution** [reply]
But I don't think it is impossible that I asked this in the wrong project.
https://github.com/withastro/roadmap/discussions/623

**Reply by JimBobSquarePants** [reply]
Yeah that’s these guys
https://github.com/lovell/sharp

**Reply by Agrarvolution** [reply]
Ah, I'm terribly sorry for bothering this project. I would have never thought to be _that_ guy who can't find the correct Github project.

**Comment by winscripter** [comment]
I would definitely love this.

Unlike JXS which is patented, JXL is completely royalty free, thanks to Google's perpetual, worldwide, royalty‑free patent license covering its official implementation.

I found the reference software (libjxl) here https://github.com/libjxl/libjxl/tree/main and by this point I definitely understand what the mass of "complexity" means here. There's a total of 109,286 lines of code.

Given that I worked in depth with multimedia before (and in fact, currently in progress of writing video codecs in C#), I believe I have enough experience to use libjxl as reference and implement JXL support for ImageSharp, though I'm not sure if JXL support is planned in the first place.

**Reply by JimBobSquarePants** [reply]
It's definitely on my TODO list. 

If you have the time and patience (you'll need it!) then please make a start as a draft so we can continuously review.

**Reply by winscripter** [reply]
I've begun working on JXL support - details are in PR #3153
